inference-engine/thirdparty/clDNN/src/gpu/loop_gpu.cpp

```diff
@@ -0,0 +1,459 @@
+    const loop_node& node;
+
+    explicit loop_gpu(const loop_node& node) : node(node) {}
+
+    static memory_impl::ptr get_outer_output_memory(const loop_inst& instance, const primitive_id& external_id) {
+        if (external_id == instance.id()) {
+            // main output
+            memory_impl& memory = instance.output_memory();
+            return (memory_impl::ptr) &memory;
+        } else {
+            // other primitives
+            const auto outputPrim = instance.get_network().get_primitive(external_id);
+            memory_impl& memory = outputPrim->output_memory();
+            return (memory_impl::ptr) &memory;
+        }
+    }
+
+    std::vector<loop::primitive_mapping> find_primitive_mappings(const primitive_id& internal_id) {
+        const auto& input_mappings = node.get_input_mappings();
+        const auto& output_mappings = node.get_output_mappings();
+
+        std::vector<loop::primitive_mapping> ret;
+        for (const auto& pm : input_mappings) {
+            if (pm.internal_id == internal_id) {
+                ret.push_back(pm);
+            }
+        }
+        for (const auto& pm : output_mappings) {
+            if (pm.internal_id == internal_id) {
+                ret.push_back(pm);
+            }
+        }
+
+        return ret;
+    }
+
+    std::vector<memory_impl::ptr> find_cropped_mem(const primitive_id& internal_id, const loop_inst& instance) {
+        const auto& input_iteration_mem = instance.input_iteration_mem;
+        for (const auto& iter_info : input_iteration_mem) {
+            if (iter_info.to_id == internal_id) {
+                return iter_info.sliced_mem;
+            }
+        }
+        const auto& output_iteration_mem = instance.output_iteration_mem;
+        for (const auto& iter_info : output_iteration_mem) {
+            if (iter_info.from_id == internal_id) {
+                return iter_info.sliced_mem;
+            }
+        }
+        return {}; // not found
+    }
+
+    void preprocess_output_memory(loop_inst& instance) {
+        auto& engine = instance.get_network().get_engine();
+        auto& output_iteration_mem = instance.output_iteration_mem;
+        auto body_network = instance.get_body_network();
+        const auto& output_primitive_map = node.get_output_mappings();
+        output_iteration_mem.reserve(output_primitive_map.size());
+        for (size_t i = 0; i < output_primitive_map.size(); ++i) {
+            const auto& primitive_map = output_primitive_map.at(i);
+            const primitive_id& external_id = primitive_map.external_id;
+            const primitive_id& internal_id = primitive_map.internal_id;
+            if (primitive_map.axis < 0) {
+                memory_impl::ptr memory = get_outer_output_memory(instance, external_id);
+                body_network->get_primitive(internal_id)->set_output_memory(*memory);
+            } else {
+                memory_impl::ptr to_mem = get_outer_output_memory(instance, external_id);
```

Comment:
to_mem/from_mem is being used by backedge so might be misleading here. I guess this to be "concat_output_mem"

Reply:
Also many namings are co-existing for the memory being used throughout iteration: iter_mem, sliced_mem, cropped_mem, etc.. Lets use one unified naming convention => For me "input_mem_concatenated" vs "input_mem_sliced" or "input_mem_concat" vs "input_mem_per_iter" .. or ..etc